foundationdb/src/error.rs

```diff
@@ -8,91 +8,167 @@
 
 //! Error types for the Fdb crate
 
-use options;
 use std;
 use std::ffi::CStr;
+use std::fmt::{self, Display};
+
+use failure::{self, Backtrace, Context, Fail};
 
 use foundationdb_sys as fdb_sys;
+use options;
+use tuple;
 
 pub(crate) fn eval(error_code: fdb_sys::fdb_error_t) -> Result<()> {
     let rust_code = error_code as i32;
     if rust_code == 0 {
         Ok(())
     } else {
-        Err(FdbError::from(error_code))
+        Err(Error::from(error_code))
     }
 }
 
-/// An error from Fdb with associated code and message
-#[derive(Debug, Fail)]
-#[fail(display = "FoundationDB error({}): {}", error_code, error_str)]
-pub struct FdbError {
-    error_code: i32,
-    error_str: &'static str,
-
+/// The Standard Error type of FoundationDB
+#[derive(Debug)]
+pub struct Error {
+    kind: Context<ErrorKind>,
     // Not all retryable error should be retried. For example, error_code=1020 transaction conflict
-    // error is always retryable (so `FdbError::is_retryable` always returns `true`), while it
+    // error is always retryable (so `Error::is_retryable` always returns `true`), while it
     // should not be retried if `TransactionOption::RetryLimit` is exceeded. So we should keep
     // track whether the error should be retried or not.
     should_retry: bool,
 }
 
+/// An error from Fdb with associated code and message
+#[derive(Debug, Fail)]
+pub enum ErrorKind {
+    /// Errors that originate from the FoundationDB layers
+    #[fail(display = "FoundationDB error({}): {}", error_code, error_str)]
+    Fdb {
+        /// The FoundationDB error code
+        error_code: i32,
+        /// The error string as defined by FoundationDB
+        error_str: &'static str,
+    },
+    /// Other errors not directly attributed to the FoundationDB layers
+    #[fail(display = "FoundationDB-RS error: {}", error)]
+    Other {
+        /// The cause of this error
+        error: failure::Error,
+    },
+    /// Encoding/Decoding errors related to Tuple
+    #[fail(display = "Internal error with tuple encoding/decoding: {}", error)]
+    Tuple {
+        /// The cause of this error
+        error: tuple::Error,
+    },
+}
+
 /// An Fdb Result type
-pub type Result<T> = std::result::Result<T, FdbError>;
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Fail for Error {
+    fn cause(&self) -> Option<&Fail> {
+        self.kind.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.kind.backtrace()
+    }
+}
 
-impl FdbError {
-    /// Converts from the raw Fdb error code into an `FdbError`
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        Display::fmt(&self.kind, f)
+    }
+}
+
+impl From<failure::Error> for Error {
+    fn from(error: failure::Error) -> Self {
+        Error::from_other(error)
+    }
+}
+
+impl Error {
```

Comment:
I'm not sure of this. As I said on previous comment, I believe user should be able to get raw `FdbError` when it is an only error type that user can get, especially for layer developers, which will want to access a API without noise. I agree that that will make user annoying when trying to use, for example `Database::transact` with `Tuple::decode`, but it could be fixed by using `failure` or `error-chain`. I tried to demonstrate an example on here:

https://github.com/yjh0502/foundationdb-rs/commit/838ecb2387b7eb57ff890489d100daaf38d7925d#diff-8fe8d202ec6a71cdcaa54f69083da070R27

Reply:
Ok, I've pulled in your changes. That was definitely more more elegant solution, I was one step away, thanks for showing how to do it.

> I'm not sure of this. As I said on previous comment, I believe user should be able to get raw FdbError when it is an only error type that user can get, especially for layer developers, which will want to access a API without noise.

We can drop the Other from this error, but I think having a common error type in the library is generally something people have grown to expect. Granted, after you're changes, and the suggestion to move to `From` on the transact interface the initial reason for this change can be dropped.

But, I think having a common type that covers both the FoundationDB errors and TupleError will be nice for us in the library. So I can drop the `Other`. I'm willing to do either at this point, but I do think this makes the library simpler to work with. It doesn't sound like either of us are strongly in favor of one direction or the other here. If it's helpful, I'll drop it from this PR, but I'm of the opinion that it does make things easier for others...

Reply:
I got your point, and I agree with your opinion. Thanks!